modules/040-control-plane-manager/docs/internal/ETCD_RECOVERY.md

```diff
@@ -5,7 +5,271 @@ Before doing this, make sure that etcd is not running. To stop etcd, remove the
 
 ## Single-master
 
-### "Failed to find database snapshot file (snap: snapshot file doesn't exist)"
+### Restoring from a backup
+
+Follow these steps to restore from a backup:
+1. Upload [etcdctl](https://github.com/etcd-io/etcd/releases) to the server (best if it has the same version as the etcd version on the server).
+
+   ```shell
+   wget "https://github.com/etcd-io/etcd/releases/download/v3.5.4/etcd-v3.5.4-linux-amd64.tar.gz"
+   tar -xzvf etcd-v3.5.4-linux-amd64.tar.gz && mv etcd-v3.5.4-linux-amd64/etcdctl /usr/local/bin/etcdctl
+   ```
+
+1. Stop etcd.
+
+   ```shell
+   mv /etc/kubernetes/manifests/etcd.yaml ~/etcd.yaml
+   ```
+
+1. Back up your files.
+
+   ```shell
+   cp -r /var/lib/etcd/member/ /var/lib/deckhouse-etcd-backup
+   ```
+
+1. Delete the data directory.
+
+   ```shell
+   rm -rf /var/lib/etcd/member/
+   ```
+
+1. Copy backup file to `~/etc-backup.snapshot`.
+
+1. Restore the etcd database.
+
+   ```shell
+   ETCDCTL_API=3 etcdctl snapshot restore ~/etc-backup.snapshot --cacert /etc/kubernetes/pki/etcd/ca.crt --cert /etc/kubernetes/pki/etcd/ca.crt \
+     --key /etc/kubernetes/pki/etcd/ca.key --endpoints https://127.0.0.1:2379/  --data-dir=/var/lib/etcd
+   ```
+
+1. Start etcd.
+
+   ```shell
+   mv ~/etcd.yaml /etc/kubernetes/manifests/etcd.yaml
+   ```
+
+## Multi-master
+
+### Complete data loss or recovery to previous state from a backup
+
+If there is a complete loss of data, perform the following steps on all nodes of the etcd cluster:
+1. Stop etcd.
+
+   ```shell
+   mv /etc/kubernetes/manifests/etcd.yaml ~/etcd.yaml
+   ```
+
+1. Back up your files.
+
+   ```shell
+   cp -r /var/lib/etcd/member/ /var/lib/deckhouse-etcd-backup
+   ```
+
+1. Delete the data directory.
+
+   ```shell
+   rm -rf /var/lib/etcd/member/
+   ```
+
+Select any node as first recovered.
+
+On another (two) nodes do the following:
+1. Stop kubelet.
+
+   ```shell
+   systemctl stop kubelet.service
+   ```
+
+1. Remove all containers.
+
+   ```shell
+   systemctl list-units --full --all | grep -q docker.service && systemctl restart docker
+   kill $(ps ax | grep containerd-shim | grep -v grep |awk '{print $1}')
```

Comment:
Do we really need to stop kubelet and restart docker?
If we'll not doing that crictl will be available on the node:

```
crictl pods --name etcd-${HOSNTAME}
```

Reply:
Yep, we should restart kubelet because on node works control-plane manager. It can be restarted and we can get state when control-plane-manager will start etcd with empty state and node can not join in the cluster